regression-test/src/androidTest/java/okhttp/regression/compare/AndroidHttpEngineTest.kt

```diff
@@ -0,0 +1,198 @@
+    context.cacheDir.resolve("httpEngine").also {
+      it.mkdirs()
+    }
+  val engine =
+    HttpEngine.Builder(context)
+      .setEnableHttp2(true)
+      .setEnableQuic(true)
+      .setEnableBrotli(true)
+      .setStoragePath(cacheDir.path)
+      .setConnectionMigrationOptions(
+        ConnectionMigrationOptions.Builder()
+          .setDefaultNetworkMigration(ConnectionMigrationOptions.MIGRATION_OPTION_ENABLED)
+          .setPathDegradationMigration(ConnectionMigrationOptions.MIGRATION_OPTION_ENABLED)
+          .setAllowNonDefaultNetworkUsage(ConnectionMigrationOptions.MIGRATION_OPTION_ENABLED)
+          .build(),
+      )
+      .setDnsOptions(
+        DnsOptions.Builder()
+          .setUseHttpStackDnsResolver(DnsOptions.DNS_OPTION_ENABLED)
+          .setStaleDns(DnsOptions.DNS_OPTION_ENABLED)
+          .setPersistHostCache(DnsOptions.DNS_OPTION_ENABLED)
+          .build(),
+      )
+      .setQuicOptions(
+        QuicOptions.Builder()
+          .addAllowedQuicHost("google.com")
+          .addAllowedQuicHost("www.google.com")
+          .build(),
+      )
+      .build()
+
+  @After
+  fun tearDown() {
+    cacheDir.deleteRecursively()
+  }
+
+  @Test
+  fun get() {
+    val executor = Executors.newCachedThreadPool()
+
+    val completableFuture = execute(engine, executor, "https://google.com/robots.txt")
+
+    try {
+      val (text, code) = completableFuture.get(10, TimeUnit.SECONDS)
+
+      assertEquals(200, code)
+      assertTrue(text.contains("Disallow"))
+    } catch (ee: ExecutionException) {
+      throw ee.cause?.cause ?: ee.cause!!
+    }
+  }
+
+  private fun execute(
+    engine: HttpEngine,
+    executor: ExecutorService,
+    url: String,
+  ): CompletableFuture<Pair<String, Int>> {
+    val completableFuture = CompletableFuture<Pair<String, Int>>()
+    val buffer = Buffer()
+    var code: Int? = null
+
+    val req =
+      engine.newUrlRequestBuilder(
+        url,
+        executor,
+        object : Callback {
+          override fun onRedirectReceived(
+            request: UrlRequest,
+            info: UrlResponseInfo,
+            newLocationUrl: String,
+          ) {
+            println("request " + info.httpStatusCode + " " + newLocationUrl)
+            request.followRedirect()
+          }
+
+          override fun onResponseStarted(
+            request: UrlRequest,
+            info: UrlResponseInfo,
+          ) {
+            println("onResponseStarted ${info.headers.asMap} ${info.negotiatedProtocol}")
+            code = info.httpStatusCode
+            request.read(ByteBuffer.allocateDirect(4096 * 8))
+          }
+
+          override fun onReadCompleted(
+            request: UrlRequest,
+            info: UrlResponseInfo,
+            byteBuffer: ByteBuffer,
+          ) {
+            println("onReadCompleted ${info.headers.asMap}")
+            byteBuffer.flip()
+            buffer.write(byteBuffer)
+            byteBuffer.clear()
+            request.read(byteBuffer)
```

Comment:
Off topic: I find this API to be completely confusing

Reply:
Yeah, agreed. I struggled to write this and I've used it before.

It is possibly optimal with zero copy transfer etc. and you can either consume like I have here as response bytes are received, or out of line as a caller requests them.